src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -326,12 +330,38 @@ inline void ShenandoahMark::mark_through_ref(T *p, ShenandoahObjToScanQueue* q,
       // and by mutator write barriers.  Assert
```

Comment:
I assume "Assert" here meant: "Assert card is dirty, instead of marking it."
Would be worth doing that. It makes sense that this card dirtying is redundant, since now old mark has been cleared before that needs to be remarked, nor a new cross-gen pointer being created here.

May be we need a ticket to get rid of this separately.

Reply:
This assert passes tier1. I'll run it through the full pipeline and see if it pops. If this code turns out to be redundant, I'd remove it on a different PR.